File: src/main/java/gregtechlite/gtlitecore/client/renderer/handler/UniversiumItemRenderer.java
package gregtechlite.gtlitecore.client.renderer.handler;

import gregtech.api.items.metaitem.MetaItem;
import gregtechlite.gtlitecore.client.model.WrappedModelGetter;
import gregtechlite.gtlitecore.client.renderer.CustomItemRenderer;
import gregtechlite.gtlitecore.client.renderer.UniversiumRenderBehavior;
import gregtechlite.gtlitecore.client.shader.CosmicShaderHelper;
import net.minecraft.client.renderer.GlStateManager;
import net.minecraft.client.renderer.block.model.IBakedModel;
import net.minecraft.client.renderer.block.model.ItemCameraTransforms;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.item.ItemStack;
import net.minecraft.world.World;
import net.minecraftforge.common.model.IModelState;
import org.lwjgl.opengl.GL11;

public class UniversiumItemRenderer extends WrappedItemRenderer
{

    public UniversiumItemRenderer(IModelState state, IBakedModel model)
    {
        super(state, model);
    }

    public UniversiumItemRenderer(IModelState state, WrappedModelGetter getter)
    {
        super(state, getter);
    }

    @Override
    public void renderItem(ItemStack stack, ItemCameraTransforms.TransformType transformType)
    {
        if (!(stack.getItem() instanceof MetaItem))
        {
            renderModel(wrapped, stack);
            return;
        }

        MetaItem<?>.MetaValueItem valueItem = ((MetaItem<?>) stack.getItem()).getItem(stack);
        UniversiumRenderBehavior urb = null;
        if (valueItem != null)
        {
            urb = (UniversiumRenderBehavior) ((CustomItemRenderer) valueItem).getRendererManager();
        }

        processLightLevel(transformType);

        if (transformType == net.minecraft.client.renderer.block.model.ItemCameraTransforms.TransformType.GUI)
        {
            renderInventory(stack, renderEntity, urb);
        }
        else
        {
            renderSimple(stack, renderEntity, urb);
        }
    }

    protected void renderSimple(ItemStack stack, EntityLivingBase player, UniversiumRenderBehavior urb)
    {
        GlStateManager.pushMatrix();
        GlStateManager.enableBlend();
        GlStateManager.blendFunc(GlStateManager.SourceFactor.SRC_ALPHA, GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA);
        GlStateManager.color(1F, 1F, 1F, 1F);

        World world = player != null ? player.world : null;
        IBakedModel model = wrapped.getOverrides().handleItemState(wrapped, stack, world, player);

        // Render base model
        renderModel(model, stack);

        // Render cosmic overlay with shader
        if (urb != null) {
            GlStateManager.disableAlpha();
            GlStateManager.depthFunc(GL11.GL_EQUAL);

            CosmicShaderHelper.cosmicOpacity = urb.getCosmicOpacity(stack);
            CosmicShaderHelper.useShader();

            renderModel(model, stack);

            CosmicShaderHelper.releaseShader();

            GlStateManager.depthFunc(GL11.GL_LEQUAL);
            GlStateManager.enableAlpha();
        }

        GlStateManager.disableBlend();
        GlStateManager.popMatrix();
    }

    protected void renderInventory(ItemStack stack, EntityLivingBase player, UniversiumRenderBehavior urb)
    {
        GlStateManager.pushMatrix();
        GlStateManager.enableBlend();
        GlStateManager.blendFunc(GlStateManager.SourceFactor.SRC_ALPHA, GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA);
        GlStateManager.disableAlpha();
        GlStateManager.disableDepth();

        World world = player != null ? player.world : null;
        IBakedModel model = wrapped.getOverrides().handleItemState(wrapped, stack, world, player);

        // Render base model
        renderModel(model, stack);

        // Render cosmic overlay with shader
        if (urb != null) {
            GlStateManager.pushMatrix();
            GlStateManager.enableBlend();
            GlStateManager.blendFunc(GlStateManager.SourceFactor.SRC_ALPHA, GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA);
            GlStateManager.disableAlpha();
            GlStateManager.disableDepth();

            GlStateManager.color(1F, 1F, 1F, 1F);
            CosmicShaderHelper.cosmicOpacity = urb.getCosmicOpacity(stack);
            CosmicShaderHelper.inventoryRender = true;
            CosmicShaderHelper.useShader();

            renderModel(model, stack);

            CosmicShaderHelper.releaseShader();
            CosmicShaderHelper.inventoryRender = false;
            GlStateManager.popMatrix();
        }

        GlStateManager.enableAlpha();
        GlStateManager.enableRescaleNormal();
        GlStateManager.enableDepth();
        GlStateManager.disableBlend();
        GlStateManager.popMatrix();
    }

    protected void processLightLevel(ItemCameraTransforms.TransformType transformType)
    {
        switch (transformType)
        {
            case GROUND:
                if (entityPos != null)
                {
                    CosmicShaderHelper.setLightFromLocation(world, entityPos);
                    return;
                }
                break;
            case THIRD_PERSON_LEFT_HAND:
            case THIRD_PERSON_RIGHT_HAND:
            case FIRST_PERSON_LEFT_HAND:
            case FIRST_PERSON_RIGHT_HAND:
            case HEAD:
                if (renderEntity != null)
                {
                    CosmicShaderHelper.setLightFromLocation(world, entityPos);
                    return;
                }
                break;
            case GUI:
                CosmicShaderHelper.setLightLevel(1.2F);
                return;
            default:
                break;
        }
        CosmicShaderHelper.setLightLevel(1.0F);
    }
}

File: src/main/kotlin/gregtechlite/gtlitecore/client/renderer/handler/item/UniversiumItemRenderer.kt
package gregtechlite.gtlitecore.client.renderer.handler.item

import gregtech.api.items.metaitem.MetaItem
import gregtechlite.gtlitecore.client.model.WrappedModelGetter
import gregtechlite.gtlitecore.client.renderer.CustomItemRenderer
import gregtechlite.gtlitecore.client.renderer.UniversiumRenderBehavior
import gregtechlite.gtlitecore.client.shader.CosmicShaderHelper
import gregtechlite.gtlitecore.client.shader.CosmicShaderHelper.releaseShader
import gregtechlite.gtlitecore.client.shader.CosmicShaderHelper.setLightLevel
import gregtechlite.gtlitecore.client.shader.CosmicShaderHelper.useShader
import net.minecraft.client.renderer.GlStateManager
import net.minecraft.client.renderer.block.model.IBakedModel
import net.minecraft.client.renderer.block.model.ItemCameraTransforms.TransformType
import net.minecraft.entity.EntityLivingBase
import net.minecraft.item.ItemStack
import net.minecraftforge.common.model.IModelState
import org.lwjgl.opengl.GL11

class UniversiumItemRenderer : WrappedItemRenderer
{
    @Suppress("unused")
    constructor(state: IModelState?, model: IBakedModel?): super(state, model)

    constructor(state: IModelState?, getter: WrappedModelGetter): super(state, getter)

    override fun renderItem(stack: ItemStack, transformType: TransformType?)
    {
        if (stack.item !is MetaItem<*>)
        {
            renderModel(wrapped!!, stack)
            return
        }

        val valueItem = (stack.item as MetaItem<*>).getItem(stack)
        var renderBehavior: UniversiumRenderBehavior? = null
        if (valueItem != null)
        {
            renderBehavior = (valueItem as CustomItemRenderer).rendererManager as UniversiumRenderBehavior
        }

        processLightLevel(transformType!!)

        if (transformType == TransformType.GUI)
        {
            renderInventory(stack, renderEntity, renderBehavior)
        }
        else
        {
            renderSimple(stack, renderEntity, renderBehavior)
        }
    }

    private fun renderSimple(stack: ItemStack, player: EntityLivingBase?, renderBehavior: UniversiumRenderBehavior?)
    {
        GlStateManager.pushMatrix()
        GlStateManager.enableBlend()
        GlStateManager.blendFunc(GlStateManager.SourceFactor.SRC_ALPHA, GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA)
        GlStateManager.color(1f, 1f, 1f, 1f)

        val world = player?.world
        val model = wrapped!!.overrides.handleItemState(wrapped!!, stack, world, player)
        renderModel(model, stack)

        if (renderBehavior != null)
        {
            GlStateManager.disableAlpha()
            GlStateManager.depthFunc(GL11.GL_EQUAL)

            CosmicShaderHelper.cosmicOpacity = renderBehavior.getCosmicOpacity(stack)
            useShader()
            renderModel(model, stack)
            releaseShader()

            GlStateManager.depthFunc(GL11.GL_LEQUAL)
            GlStateManager.enableAlpha()
        }

        GlStateManager.disableBlend()
        GlStateManager.popMatrix()
    }

    private fun renderInventory(stack: ItemStack, player: EntityLivingBase?, renderBehavior: UniversiumRenderBehavior?)
    {
        GlStateManager.pushMatrix()
        GlStateManager.enableBlend()
        GlStateManager.blendFunc(GlStateManager.SourceFactor.SRC_ALPHA, GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA)
        GlStateManager.disableAlpha()
        GlStateManager.disableDepth()

        val world = player?.world
        val model = wrapped!!.overrides.handleItemState(wrapped!!, stack, world, player)
        renderModel(model, stack)

        if (renderBehavior != null)
        {
            GlStateManager.pushMatrix()
            GlStateManager.enableBlend()
            GlStateManager.blendFunc(GlStateManager.SourceFactor.SRC_ALPHA, GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA)
            GlStateManager.disableAlpha()
            GlStateManager.disableDepth()
            GlStateManager.color(1f, 1f, 1f, 1f)

            CosmicShaderHelper.cosmicOpacity = renderBehavior.getCosmicOpacity(stack)
            CosmicShaderHelper.inventoryRender = true
            useShader()
            renderModel(model, stack)
            releaseShader()
            CosmicShaderHelper.inventoryRender = false

            GlStateManager.popMatrix()
        }
        GlStateManager.enableAlpha()
        GlStateManager.enableRescaleNormal()
        GlStateManager.enableDepth()
        GlStateManager.disableBlend()
        GlStateManager.popMatrix()
    }

    private fun processLightLevel(transformType: TransformType)
    {
        when (transformType)
        {
            TransformType.GROUND -> {
                if (entityPos != null)
                {
                    CosmicShaderHelper.setLightFromLocation(world, entityPos!!)
                    return
                }
            }

            TransformType.THIRD_PERSON_LEFT_HAND,
            TransformType.THIRD_PERSON_RIGHT_HAND,
            TransformType.FIRST_PERSON_LEFT_HAND,
            TransformType.FIRST_PERSON_RIGHT_HAND,
            TransformType.HEAD -> {
                if (renderEntity != null)
                {
                    CosmicShaderHelper.setLightFromLocation(world, entityPos!!)
                    return
                }
            }
            TransformType.GUI ->
            {
                setLightLevel(1.2f)
                return
            }
            else -> {}
        }
        setLightLevel(1.0f)
    }
}